File: apps/ios/ExpCore/Sources/Domain/IssueRelationType.swift
import Foundation

/// The four canonical relation types (`issue_relations.type`, EXP-736), the
/// hand-maintained mirror of `packages/domain-contract/contract.json`
/// `issueRelationType` — locked to the generated constants by
/// IssueRelationContractTests, exactly like `IssueStatus`.
///
/// Every type has ONE canonical direction, stored on the row: the row's
/// `issue_id` side is the FORWARD side and `related_issue_id` the INVERSE one.
/// `blocks` = the issue blocks the related one, `parent` = the issue is the
/// parent of the related one, `duplicate` = the issue IS the duplicate (the
/// related one is canonical, mirroring `issues.duplicate_of_id`), `related` is
/// symmetric and reads the same from both sides.
public enum IssueRelationType: String, CaseIterable, Codable, Identifiable, Sendable {
    case blocks
    case parent
    case duplicate
    case related

    public var id: String { rawValue }

    /// The label for ONE side of the row: `false` reads it from the row's own
    /// issue (forward), `true` from the related issue. Byte-identical to
    /// contract `forwardLabels` / `inverseLabels` on all four clients.
    public func label(inverse: Bool) -> String {
        switch self {
        case .blocks: inverse ? "blocked by" : "blocks"
        case .parent: inverse ? "sub-issue of" : "parent of"
        case .duplicate: inverse ? "duplicated by" : "duplicate of"
        case .related: "related to"
        }
    }

    /// Tolerant wire decoder — an unknown type (an older client meeting a
    /// newer server) yields nil and the row is simply not rendered.
    public static func from(_ wire: String?) -> IssueRelationType? {
        guard let wire else { return nil }
        return IssueRelationType(rawValue: wire)
    }
}

/// One entry of the "Add relation" picker: the wire type plus which side the
/// caller's issue ends up on. Order and wording are shared across the four
/// clients — Parent of, Sub-issue of, Blocking, Blocked by, Duplicate of,
/// Related to.
public struct RelationPick: Identifiable, Sendable, Equatable {
    public let type: IssueRelationType
    /// True = the caller's issue is the RELATED side of the stored row.
    public let inverse: Bool
    public let title: String

    public init(type: IssueRelationType, inverse: Bool, title: String) {
        self.type = type
        self.inverse = inverse
        self.title = title
    }

    public var id: String { "\(type.rawValue)-\(inverse)" }

    public static let all: [RelationPick] = [
        RelationPick(type: .parent, inverse: false, title: "Parent of"),
        RelationPick(type: .parent, inverse: true, title: "Sub-issue of"),
        RelationPick(type: .blocks, inverse: false, title: "Blocking"),
        RelationPick(type: .blocks, inverse: true, title: "Blocked by"),
        RelationPick(type: .duplicate, inverse: false, title: "Duplicate of"),
        RelationPick(type: .related, inverse: false, title: "Related to"),
    ]

    /// Render order for a relation row: its (type, side) position in `all`.
    /// A side with no picker entry (`duplicated by`) sorts after every pick.
    public static func order(type: IssueRelationType, inverse: Bool) -> Int {
        all.firstIndex { $0.type == type && $0.inverse == inverse } ?? all.count
    }
}

extension IssueRelationType {
    /// The activity phrase for a `relation_added` / `relation_removed` event,
    /// byte-identical across web, desktop, iOS and Android. `inverse` is the
    /// event payload's `direction`, so each issue's own row reads from its own
    /// side.
    public static func eventPhrase(
        type: IssueRelationType,
        inverse: Bool,
        identifier: String,
        removed: Bool
    ) -> String {
        if type == .related {
            return removed ? "removed related issue \(identifier)" : "added related issue \(identifier)"
        }
        let label = type.label(inverse: inverse)
        return removed ? "no longer \(label) \(identifier)" : "marked as \(label) \(identifier)"
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/IssueRelationType.kt
package com.exponential.app.domain

import androidx.compose.ui.graphics.vector.ImageVector
import com.exponential.app.ui.icons.ExpIcons

/**
 * The four issue relation kinds (EXP-736). A stored row is always the
 * CANONICAL direction — `issue_id` blocks / is the parent of / is the
 * duplicate of `related_issue_id`; `related` is symmetric — so each of the two
 * issues renders the SAME row under a different label. [label] picks the side:
 * `inverse = true` is the target issue's reading of it.
 *
 * Entry order and both label lists are locked to
 * DomainContract.issueRelationType* by DomainContractLockTest.
 */
enum class IssueRelationType(
    val wire: String,
    val forwardLabel: String,
    val inverseLabel: String,
) {
    Blocks(DomainContract.issueRelationTypeBlocks, "blocks", "blocked by"),
    Parent(DomainContract.issueRelationTypeParent, "parent of", "sub-issue of"),
    Duplicate(DomainContract.issueRelationTypeDuplicate, "duplicate of", "duplicated by"),
    Related(DomainContract.issueRelationTypeRelated, "related to", "related to");

    fun label(inverse: Boolean): String = if (inverse) inverseLabel else forwardLabel

    companion object {
        /** Null for an unknown value from a newer server (the row degrades to
         *  a plain link rather than mislabeling itself). */
        fun fromWire(value: String?): IssueRelationType? = entries.firstOrNull { it.wire == value }
    }
}

// The `direction` values a relation event payload carries — which SIDE of the
// edge the event was recorded for, hence which label its phrase reads.
const val RELATION_DIRECTION_FORWARD = "forward"
const val RELATION_DIRECTION_INVERSE = "inverse"

/**
 * One entry of the "Add relation" picker: a human title over the (type,
 * inverse) pair the create mutation sends. Ordering and wording are shared by
 * all four clients.
 */
data class RelationPick(
    val title: String,
    val type: IssueRelationType,
    val inverse: Boolean,
) {
    val icon: ImageVector
        get() = when {
            type == IssueRelationType.Parent && !inverse -> ExpIcons.relationParent
            type == IssueRelationType.Parent -> ExpIcons.relationSubIssue
            type == IssueRelationType.Blocks && !inverse -> ExpIcons.relationBlocks
            type == IssueRelationType.Blocks -> ExpIcons.relationBlockedBy
            type == IssueRelationType.Duplicate -> ExpIcons.relationDuplicate
            else -> ExpIcons.relationRelated
        }
}

/** The six picks, in the order every client offers them. */
val relationPicks: List<RelationPick> = listOf(
    RelationPick("Parent of", IssueRelationType.Parent, inverse = false),
    RelationPick("Sub-issue of", IssueRelationType.Parent, inverse = true),
    RelationPick("Blocking", IssueRelationType.Blocks, inverse = false),
    RelationPick("Blocked by", IssueRelationType.Blocks, inverse = true),
    RelationPick("Duplicate of", IssueRelationType.Duplicate, inverse = false),
    RelationPick("Related to", IssueRelationType.Related, inverse = false),
)

/** Where a (type, side) pair sorts in a relation list — the picker's order. */
fun relationSortKey(type: IssueRelationType?, inverse: Boolean): Int {
    val index = relationPicks.indexOfFirst { it.type == type && it.inverse == inverse }
    return if (index >= 0) index else relationPicks.size
}

/** The glyph one SIDE of a relation reads as (the picker's, per side). */
fun relationIcon(type: IssueRelationType?, inverse: Boolean): ImageVector =
    relationPicks.firstOrNull { it.type == type && it.inverse == inverse }?.icon
        ?: ExpIcons.relationRelated

/**
 * The phrase a `relation_added` / `relation_removed` timeline event reads,
 * byte-identical on all four clients: `related` names the act ("added related
 * issue EXP-12"), every other type names the resulting state ("marked as
 * blocked by EXP-3").
 *
 * A thin payload never drops the row — it degrades exactly as the web
 * `relationEventPhrase` does, because an old row or a hard-deleted counterpart
 * must still read as something: an unknown/missing [type] reads as the
 * symmetric `related`, and a missing [identifier] is named "an issue".
 */
fun relationEventPhrase(
    added: Boolean,
    type: String?,
    identifier: String?,
    direction: String?,
): String {
    val named = identifier?.takeIf { it.isNotBlank() } ?: "an issue"
    val relation = IssueRelationType.fromWire(type)
    if (relation == null || relation == IssueRelationType.Related) {
        return if (added) "added related issue $named" else "removed related issue $named"
    }
    val label = relation.label(inverse = direction == RELATION_DIRECTION_INVERSE)
    return if (added) "marked as $label $named" else "no longer $label $named"
}
